alu_u32/src/com/stark.rs

```diff
@@ -0,0 +1,46 @@
+use super::columns::Com32Cols;
+use super::Com32Chip;
+use core::borrow::Borrow;
+
+use crate::com::columns::NUM_COM_COLS;
+use p3_air::{Air, AirBuilder, BaseAir};
+use p3_field::AbstractField;
+use p3_matrix::MatrixRowSlices;
+
+impl<F> BaseAir<F> for Com32Chip {
+    fn width(&self) -> usize {
+        NUM_COM_COLS
+    }
+}
+
+impl<F, AB> Air<AB> for Com32Chip
+where
+    F: AbstractField,
+    AB: AirBuilder<F = F>,
+{
+    fn eval(&self, builder: &mut AB) {
+        let main = builder.main();
+        let local: &Com32Cols<AB::Var> = main.row_slice(0).borrow();
+
+        // Check if the first two operand values are equal, in case we're doing a conditional branch.
+        // (when is_imm == 1, the second read value is guaranteed to be an immediate value)
+        builder.assert_eq(
+            local.diff,
+            local
+                .input_1
+                .into_iter()
+                .zip(local.input_2)
+                .map(|(a, b)| (a - b) * (a - b))
+                .sum::<AB::Expr>(),
+        );
+        builder.assert_bool(local.not_equal);
+        builder.assert_eq(local.not_equal, local.diff * local.diff_inv);
+
+        builder.assert_bool(local.is_ne);
+        builder.assert_bool(local.is_eq);
+        builder.assert_bool(local.is_ne + local.is_eq);
+
+        builder.when(local.is_ne).assert_one(local.not_equal);
+        builder.when(local.is_eq).assert_zero(local.not_equal);
```

Comment:
I think these two lines could be removed as `assert_bool(local.is_ne + local.is_eq)` above should also enforce this

Reply:
True. It seems these are stronger than `assert_bool(local.is_ne + local.is_eq)` though? I've removed that line instead.

Reply:
Oic, with the output check it's fine. I've removed the one line as you suggested.

Reply:
Hm...actually the output check is still checking the two together rather than individually. So I do think it's more thorough to keep the 2 lines. Lmk what you think.

Reply:
Whoops I misread the two lines, thought it was meant to enforce that `is_ne` and `is_eq` are mutually exclusive.

After re-reading, I'm not quite clear on the intent of these two lines. I thought `is_ne`, `is_eq` were indicating what instruction was invoked, but not indicating the result, so not directly tied to `not_equal` which I understood to be the result of the equality test. Was I interpreting it right?

Reply:
Oh yes you were right, I meant to verify that the respective instructions are outputing the correct result. So it should have been something like:
`when ne assert result = not_equal` `when eq assert result = 1 - not_equal`. Sorry about the confusion.